File: ws-client.py
import websocket
import os
import json
import random
import time

url = f"wss://app.rvrb.one/ws-bot?apiKey={os.environ.get('apikey')}"
channelId = None
password = os.environ.get('password', None)

ws = None
latency = 0
reconnect = True
joinId = None

# Attempt to join my channel
def join():
    global joinId
    joinId = round(random.random() * 100) # Provide an ID to get a response
    joinRequest = {
        "method": "join",
        "params": {
            "channelId": channelId
        },
        "id": joinId
    }
    if password:
        joinRequest["params"]["password"] = password
    ws.send(json.dumps(joinRequest))

# Event handlers for the WebSocket connection
# These are called when the server sends a message
# with a method that matches the key
def keepAwake(data):
    global latency
    # Keep awake is like a ping but also used to measure latency
    latency = data["params"]["latency"]
    print(f"Latency: {latency}ms")
    # Send a stayAwake message back to the server
    # If the server doesn't receive a stayAwake message 3 times in a row
    # The server will close the connection
    ws.send(json.dumps({
        "jsonrpc": "2.0",
        "method": "stayAwake",
        "params": {
            "date": int(time.time() * 1000)
        }
    }))

def ready(data):
    global channelId
    if data.params["channelId"]:
        channelId = data.params["channelId"]
    join() # Server sends ready when it's ready to receive commands

def pushChannelMessage(data):
    print("Received chat message", data["params"])

def pushNotification(data):
    print("Received notification", data["params"]) # Notification from the server

def updateChannel(data):
    print("Received channel update", data["params"]) # Channel name or description change

def updateChannelUsers(data):
    print("Received channel users update", data["params"]) # Users join or leave

def updateUser(data):
    print("Received user update", data["params"]) # User changes name or avatar, etc.

def updateChannelDjs(data):
    print("Received channel djs update", data["params"]) # DJs change

def updateChannelMeter(data):
    print("Received channel meter update", data["params"]) # Users vote

def updateChannelUserStatus(data):
    print("Received channel user status update", data["params"]) # User AFK, active, etc.

def leaveChannel(data):
    print("Received leave channel", data["params"]) # Command to leave channel
    global reconnect
    reconnect = False
    ws.close()

def playChannelTrack(data):
    print("Received play channel track", data["params"]) # Track starts playing

def pauseChannelTrack(data):
    print("Received pause channel track", data["params"]) # Track paused

eventHandlers = {
    "keepAwake": keepAwake,
    "ready": ready,
    "pushChannelMessage": pushChannelMessage,
    "pushNotification": pushNotification,
    "updateChannel": updateChannel,
    "updateChannelUsers": updateChannelUsers,
    "updateUser": updateUser,
    "updateChannelDjs": updateChannelDjs,
    "updateChannelMeter": updateChannelMeter,
    "updateChannelUserStatus": updateChannelUserStatus,
    "leaveChannel": leaveChannel,
    "playChannelTrack": playChannelTrack,
    "pauseChannelTrack": pauseChannelTrack
}

def onMessage(ws, message):
    data = json.loads(message)
    print(f"Received message: {message}")
    # wait for the ready message before sending anything else
    if 'method' in data and data['method'] in eventHandlers:
        eventHandlers[data['method']](data)
    elif 'id' in data and data['id'] == joinId:
        # was join a success?
        if 'error' in data:
            print(f"Error joining channel: {data['error']['message']}")
        else:
            print(f"Joined channel {channelId}")

def onPing(ws, message):
    print("Received ping from server")
    ws.pong()

def onPong(ws, message):
    print("Received pong from server")
    ws.ping()

def onOpen(ws):
    print("Connected to server")

def onClose(ws):
    print("Disconnected from server")
    # reconnect?
    if reconnect:
        ws.run_forever()

def onError(ws, error):
    print(f"WebSocket error: {error}")

def connect():
    global ws
    # attempt to connect to the WebSocket server
    ws = websocket.WebSocketApp(url, on_open=onOpen, on_message=onMessage, on_error=onError, on_close=onClose)
    ws.on_ping = onPing
    ws.on_pong = onPong
    ws.run_forever()

connect()

File: ws-client.js
const WebSocket = require('ws')

// const url = `wss://app.rvrb.one/ws-bot?apiKey=${process.env.apikey}`
const url = `wss://app.rvrb.one/ws-bot?apiKey=${process.env.apikey}`

let channelId = null
const password = process.env.password || null

let ws = null
let latency = 0
const reconnect = true
let joinId = null

// attempt to join my channel
const join = () => {
  joinId = Math.round(Math.random() * 100) // provide an id to get a response
  const joinRequest = {
    method: 'join',
    params: {
      channelId
    },
    id: joinId
  }
  if (password) {
    joinRequest.params.password = password
  }
  ws.send(JSON.stringify(joinRequest))
}

// event handlers for the WebSocket connection
// these are called when the server sends a message
// with a method that matches the key
const eventHandlers = {
  keepAwake: (data) => {
    // keep awake is like a ping but also used to measure latency
    latency = data.params.latency
    console.log(`Latency: ${latency}ms`)
    // send a stayAwake message back to the server
    // if the server doesn't receive a stayAwake message 3 times in a row
    // the server will close the connection
    ws.send(JSON.stringify({
      jsonrpc: '2.0',
      method: 'stayAwake',
      params: {
        date: Date.now()
      }
    }))
  },
  ready: (data) => {
    if ('channelId' in data.params) {
      channelId = data.params.channelId
      join() // server sends ready when it's ready to receive commands
    }
  },
  pushChannelMessage: (data) => {
    // received a chat message from the channel
    // this is where you would handle commands
    console.log(`Received message from channel: ${data.params.message}`)
  },
  pushNotification: (data) => {
    // received a notification from the channel
    // this is where you would handle notifications
    console.log(`Received notification from channel: ${data.params.message}`)
  },
  updateChannel: (data) => {
    // received an update to the channel
    // this is where you would handle updates
    console.log(`Received update from channel: ${data.params.message}`)
  },
  updateChannelUsers: (data) => {
    // received an update to the channel users (join, leave, etc)
    // this is where you would handle updates
    console.log(`Received update from channel users: ${data.params.message}`)
  },
  updateChannelUser: (data) => {
    // received an update to a channel user (change display name, etc)
    // this is where you would handle updates
    console.log(`Received update from channel user: ${data.params.message}`)
  },
  updateChannelDjs: (data) => {
    // received an update to the channel djs (add, remove, etc)
    // this is where you would handle updates
    console.log(`Received update from channel djs: ${data.params.message}`)
  },
  updateChannelMeter: (data) => {
    // received a voting update
    // this is where you would handle updates
    console.log(`Received update from channel meter: ${data.params.message}`)
  },
  updateChannelUserStatus: (data) => {
    // received an update to a channel user's status (mute, block, afk, typing)
    // this is where you would handle updates
    console.log(`Received update from channel user status: ${data.params.message}`)
  },
  leaveChannel: (data) => {
    // received a leave channel message
    // this command instructs the bot to leave the channel, nicely
    // this is where you would handle updates
    console.log(`Received leave channel message: ${data.params.message}`)
  },
  playChannelTrack: (data) => {
    // received a play track message
    // this is sent when a new track is playing
    console.log(`Received play track message: ${data.params.message}`)
  },
  pauseChannelTrack: (data) => {
    // received a pause track message
    // this is sent when a track is paused
    console.log(`Received pause track message: ${data.params.message}`)
  }
}

const connect = () => {
  // attempt to connect to the WebSocket server
  ws = new WebSocket(url)

  ws.on('open', () => {
    console.log('Connected to server')
  })

  ws.on('close', () => {
    console.log('Disconnected from server')
    // reconnect?
    if (reconnect) {
      setTimeout(() => {
        connect()
      }, 1000)
    }
  })

  ws.on('error', (error) => {
    console.error(`WebSocket error: ${error}`)
  })

  ws.on('message', (data) => {
    console.log(`Received message: ${data}`)
    try {
      data = JSON.parse(data)
    } catch (e) {
      console.error(`Error parsing JSON: ${e}`)
      return
    }
    // wait for the ready message before sending anything else
    if (data.method in eventHandlers) {
      eventHandlers[data.method](data)
    } else if (data.id === joinId) {
      // was join a success?
      if (data.error) {
        console.error(`Error joining channel: ${data.error.message}`)
      } else {
        console.log(`Joined channel ${channelId}`)
      }
    }
  })

  ws.on('ping', () => {
    console.log('Received ping from server')
    ws.pong()
  })

  ws.on('pong', () => {
    console.log('Received pong from server')
    ws.ping()
  })
}

connect()
